clients/src/main/java/org/apache/kafka/common/requests/JoinGroupRequest.java

```diff
@@ -60,14 +62,40 @@ public ByteBuffer metadata() {
         }
     }
 
+    // v0 constructor
+    @Deprecated
```

Comment:
This is the pattern we've done elsewhere (see e.g. OffsetCommitRequest), but I'm frankly not a fan of it and it seems to be used inconsistently. I'm also unclear about what we consider a public API. In other cases, we've tried harder to keep compatibility as long as the class wasn't in an "internals" package, but again we're inconsistent.